pkg/apis/workflow/v1alpha1/workflow_types.go

```diff
@@ -840,6 +843,17 @@ type Artifact struct {
 	FromExpression string `json:"fromExpression,omitempty" protobuf:"bytes,11,opt,name=fromExpression"`
 }
 
+// ValidatePath verifies that the artifact path is valid.
+func (art Artifact) ValidatePath() error {
+	if art.Path == "" {
+		return argoerrs.InternalErrorf("Artifact %s did not specify a path", art.Name)
+	}
+	if strings.Contains(art.Path, "..") {
```

Comment:
I think this check should be more robust. As far as I can tell, `a..a` is a valid filename in *nix.

Will look into whether golang has a library to detect directory traversal...

Reply:
I guess once could use [`filepath.Clean`](https://pkg.go.dev/path/filepath#Clean) and then check that the generated path begins with the full root path of the artifacts directory.

Reply:
One option would be to fully resolve the path with `path.Clean` and make sure the final path is within some safe boundary, but I'm not sure what the safe boundary would be. Checking for `../` doesn't seem to be sufficient either since `a..` is a valid file name.

Reply:
Here's one idea: 

```go
func isSafePath(pathToTest string) bool {
	// Convert the given path to an absolute path by prepending "/" (not cross-platform?)
	cleanPath := path.Clean("/" + pathToTest)
	// Assert that the given path is no different from the clean path (i.e. no traversal)
	return cleanPath == "/" + pathToTest
}

func main() {
	fmt.Println(isSafePath("../whatever"))
	fmt.Println(isSafePath("whatever"))
}
```

Downsides:
* probably not cross-platform
* doesn't allow directory traversal _within_ the safe area (which would be nice, but probably not must-have)

Reply:
Made this function more robust in https://github.com/argoproj/argo-workflows/pull/7187/commits/6fed6857743cd4abeef3951fe4e34485162ea5f3. See test cases for some clear examples of how it behaves.